File: ios/Runner/Transfer/KarlshareProtocol.swift
import Foundation

/// Karlshare binary wire protocol (spec Section 7.2) — Swift mirror of the
/// Android `KarlshareProtocol.kt`. All multi-byte integers are big-endian
/// (network byte order) so the two platforms interop on the byte level.
///
/// Message types (one-byte tag):
///   0x01 HANDSHAKE   [VERSION:2][DEVICE_ID:16][PUBLIC_KEY:32][CAPABILITIES:1]
///   0x02 FILE_HEADER [FILE_ID:16][NAME_LEN:2][NAME][SIZE:8][MIME_LEN:1][MIME][CHECKSUM:32]
///   0x03 CHUNK       [FILE_ID:16][CHUNK_INDEX:4][CHUNK_SIZE:4][DATA]
///   0x04 ACK         [FILE_ID:16][CHUNK_INDEX:4][STATUS:1]
///   0x05 CANCEL      [FILE_ID:16]
///
/// As on Android, the HANDSHAKE tag is omitted on the wire — both peers know
/// the first frame on a fresh socket is the handshake.
enum KarlshareProtocol {

    static let version: UInt16 = 2
    static let transferPort: UInt16 = 8988
    static let defaultChunkSize: Int = 256 * 1024
    static let parallelChunks: Int = 4

    enum Tag {
        static let handshake: UInt8 = 0x01
        static let fileHeader: UInt8 = 0x02
        static let chunk: UInt8 = 0x03
        static let ack: UInt8 = 0x04
        static let cancel: UInt8 = 0x05
    }

    enum AckStatus {
        static let ok: UInt8 = 0x00
        static let retry: UInt8 = 0x01
        static let fail: UInt8 = 0x02
    }

    /// Capability bits packed into the 1-byte CAPABILITIES field.
    enum Capabilities {
        static let supportsResume = 1 << 0
        static let supportsParallelChunks = 1 << 1
        static let supportsTLS = 1 << 2
    }

    enum ProtocolError: Error {
        case eof
        case malformed(String)
    }

    // MARK: - Models

    struct Handshake {
        let version: UInt16
        let deviceId: UUID
        /// SHA-256 of the peer's TLS cert SubjectPublicKeyInfo (32 bytes).
        let publicKey: Data
        let capabilities: Int
    }

    struct FileHeader {
        let fileId: UUID
        let name: String
        let size: UInt64
        let mime: String
        let checksum: Data // SHA-256, 32 bytes
    }

    struct Chunk {
        let fileId: UUID
        let index: UInt32
        let data: Data
    }

    struct Ack {
        let fileId: UUID
        let index: UInt32
        let status: UInt8
    }

    enum Frame {
        case header(FileHeader)
        case chunk(Chunk)
        case ack(Ack)
        case cancel(UUID)
    }

    /// Reads exactly `n` bytes, or throws `.eof` if the stream closed first.
    typealias Reader = (_ n: Int) async throws -> Data

    // MARK: - Encoding

    static func encodeHandshake(_ h: Handshake) -> Data {
        var out = Data()
        out.appendBE(h.version)
        out.append(uuidBytes(h.deviceId))
        precondition(h.publicKey.count == 32, "publicKey must be 32 bytes")
        out.append(h.publicKey)
        out.append(UInt8(h.capabilities & 0xFF))
        return out
    }

    static func encodeFileHeader(_ h: FileHeader) -> Data {
        var out = Data()
        out.append(Tag.fileHeader)
        out.append(uuidBytes(h.fileId))
        let nameBytes = Data(h.name.utf8)
        precondition(nameBytes.count <= 0xFFFF, "file name too long")
        out.appendBE(UInt16(nameBytes.count))
        out.append(nameBytes)
        out.appendBE(h.size)
        let mimeBytes = Data(h.mime.utf8)
        precondition(mimeBytes.count <= 0xFF, "mime too long")
        out.append(UInt8(mimeBytes.count))
        out.append(mimeBytes)
        precondition(h.checksum.count == 32, "checksum must be 32 bytes")
        out.append(h.checksum)
        return out
    }

    static func encodeChunk(_ c: Chunk) -> Data {
        var out = Data()
        out.append(Tag.chunk)
        out.append(uuidBytes(c.fileId))
        out.appendBE(c.index)
        out.appendBE(UInt32(c.data.count))
        out.append(c.data)
        return out
    }

    static func encodeAck(_ a: Ack) -> Data {
        var out = Data()
        out.append(Tag.ack)
        out.append(uuidBytes(a.fileId))
        out.appendBE(a.index)
        out.append(a.status)
        return out
    }

    static func encodeCancel(_ fileId: UUID) -> Data {
        var out = Data()
        out.append(Tag.cancel)
        out.append(uuidBytes(fileId))
        return out
    }

    // MARK: - Decoding

    static func decodeHandshake(_ read: Reader) async throws -> Handshake {
        let version = try await readBE(UInt16.self, read)
        let deviceId = try await readUUID(read)
        let publicKey = try await read(32)
        let cap = try await read(1)
        return Handshake(
            version: version,
            deviceId: deviceId,
            publicKey: publicKey,
            capabilities: Int(cap[cap.startIndex])
        )
    }

    /// Reads one tag-prefixed frame. Returns nil on a clean EOF.
    static func decodeFrame(_ read: Reader) async throws -> Frame? {
        let tagData: Data
        do {
            tagData = try await read(1)
        } catch ProtocolError.eof {
            return nil
        }
        let tag = tagData[tagData.startIndex]
        switch tag {
        case Tag.fileHeader:
            return .header(try await decodeFileHeaderBody(read))
        case Tag.chunk:
            return .chunk(try await decodeChunkBody(read))
        case Tag.ack:
            return .ack(try await decodeAckBody(read))
        case Tag.cancel:
            return .cancel(try await readUUID(read))
        default:
            throw ProtocolError.malformed(String(format: "unknown frame tag 0x%02x", tag))
        }
    }

    private static func decodeFileHeaderBody(_ read: Reader) async throws -> FileHeader {
        let fileId = try await readUUID(read)
        let nameLen = Int(try await readBE(UInt16.self, read))
        let name = String(decoding: try await read(nameLen), as: UTF8.self)
        let size = try await readBE(UInt64.self, read)
        let mimeLen = Int(try await read(1)[0])
        let mime = String(decoding: try await read(mimeLen), as: UTF8.self)
        let checksum = try await read(32)
        return FileHeader(fileId: fileId, name: name, size: size, mime: mime, checksum: checksum)
    }

    private static func decodeChunkBody(_ read: Reader) async throws -> Chunk {
        let fileId = try await readUUID(read)
        let index = try await readBE(UInt32.self, read)
        let size = Int(try await readBE(UInt32.self, read))
        guard size >= 0, size <= 8 * 1024 * 1024 else {
            throw ProtocolError.malformed("implausible chunk size: \(size)")
        }
        let data = try await read(size)
        return Chunk(fileId: fileId, index: index, data: data)
    }

    private static func decodeAckBody(_ read: Reader) async throws -> Ack {
        let fileId = try await readUUID(read)
        let index = try await readBE(UInt32.self, read)
        let status = try await read(1)[0]
        return Ack(fileId: fileId, index: index, status: status)
    }

    private static func readUUID(_ read: Reader) async throws -> UUID {
        let bytes = try await read(16)
        guard let uuid = uuidFromBytes(bytes) else {
            throw ProtocolError.malformed("bad UUID bytes")
        }
        return uuid
    }

    private static func readBE<T: FixedWidthInteger>(_ type: T.Type, _ read: Reader) async throws -> T {
        let bytes = try await read(MemoryLayout<T>.size)
        var value: T = 0
        for b in bytes { value = (value << 8) | T(b) }
        return value
    }

    // MARK: - UUID <-> bytes (big-endian, matches Java UUID MSB/LSB layout)

    static func uuidBytes(_ uuid: UUID) -> Data {
        var u = uuid.uuid
        return withUnsafeBytes(of: &u) { Data($0) }
    }

    static func uuidFromBytes(_ data: Data) -> UUID? {
        guard data.count == 16 else { return nil }
        let b = [UInt8](data)
        let t: uuid_t = (b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7],
                         b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15])
        return UUID(uuid: t)
    }
}

// MARK: - Big-endian append helpers

private extension Data {
    mutating func appendBE<T: FixedWidthInteger>(_ value: T) {
        let big = value.bigEndian
        withUnsafeBytes(of: big) { append(contentsOf: $0) }
    }
}

File: android/app/src/main/kotlin/com/karlshare/karlshare/transfer/KarlshareProtocol.kt
package com.karlshare.karlshare.transfer

import java.io.DataInputStream
import java.io.DataOutputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.UUID

/**
 * Karlshare binary wire protocol (spec Section 7.2).
 *
 * All multi-byte integers are big-endian (network byte order).
 *
 * Message types (one-byte tag):
 *   0x01 HANDSHAKE      [VERSION:2][DEVICE_ID:16][PUBLIC_KEY:32][CAPABILITIES:1]
 *   0x02 FILE_HEADER    [FILE_ID:16][NAME_LEN:2][NAME:variable]
 *                       [SIZE:8][MIME_LEN:1][MIME:variable][CHECKSUM:32]
 *   0x03 CHUNK          [FILE_ID:16][CHUNK_INDEX:4][CHUNK_SIZE:4][DATA:variable]
 *   0x04 ACK            [FILE_ID:16][CHUNK_INDEX:4][STATUS:1]
 *   0x05 CANCEL         [FILE_ID:16]
 *
 * The HANDSHAKE tag is omitted from the on-wire bytes for the handshake itself
 * — both peers know the first frame on a fresh socket is the handshake.
 */
object KarlshareProtocol {

    /**
     * Bumped to 2 in the TLS rollout. v1 spoke plaintext TCP; v2 mandates
     * TLS 1.3 + cert-fingerprint binding on the handshake's PUBLIC_KEY
     * field. v2 peers will never connect to v1 peers (TLS handshake would
     * fail before the Karlshare handshake), which is the intended behavior.
     */
    const val PROTOCOL_VERSION: Short = 2
    const val TRANSFER_PORT: Int = 8988
    const val DEFAULT_CHUNK_SIZE: Int = 256 * 1024
    const val PARALLEL_CHUNKS: Int = 4

    const val TYPE_HANDSHAKE: Byte = 0x01
    const val TYPE_FILE_HEADER: Byte = 0x02
    const val TYPE_CHUNK: Byte = 0x03
    const val TYPE_ACK: Byte = 0x04
    const val TYPE_CANCEL: Byte = 0x05

    const val ACK_OK: Byte = 0x00
    const val ACK_RETRY: Byte = 0x01
    const val ACK_FAIL: Byte = 0x02

    /** Capability bits packed into the 1-byte CAPABILITIES field. */
    object Capabilities {
        const val SUPPORTS_RESUME: Int = 1 shl 0
        const val SUPPORTS_PARALLEL_CHUNKS: Int = 1 shl 1
        const val SUPPORTS_TLS: Int = 1 shl 2 // on by default since v2
    }

    data class Handshake(
        val version: Short,
        val deviceId: UUID,
        /**
         * Since v2 this is the SHA-256 of the peer's TLS cert
         * SubjectPublicKeyInfo (32 bytes). After the TLS handshake completes
         * we cross-check the negotiated cert against this value — that's the
         * binding that makes our self-signed-cert TOFU model MITM-resistant.
         */
        val publicKey: ByteArray,
        val capabilities: Int,
    ) {
        init {
            require(publicKey.size == 32) { "publicKey must be 32 bytes" }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Handshake) return false
            return version == other.version &&
                deviceId == other.deviceId &&
                publicKey.contentEquals(other.publicKey) &&
                capabilities == other.capabilities
        }

        override fun hashCode(): Int {
            var result = version.toInt()
            result = 31 * result + deviceId.hashCode()
            result = 31 * result + publicKey.contentHashCode()
            result = 31 * result + capabilities
            return result
        }
    }

    data class FileHeader(
        val fileId: UUID,
        val name: String,
        val size: Long,
        val mime: String,
        val checksum: ByteArray, // SHA-256
    ) {
        init {
            require(checksum.size == 32) { "checksum must be 32 bytes (SHA-256)" }
            require(name.toByteArray(StandardCharsets.UTF_8).size <= 0xFFFF) {
                "file name too long"
            }
            require(mime.toByteArray(StandardCharsets.UTF_8).size <= 0xFF) {
                "mime type too long"
            }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is FileHeader) return false
            return fileId == other.fileId &&
                name == other.name &&
                size == other.size &&
                mime == other.mime &&
                checksum.contentEquals(other.checksum)
        }

        override fun hashCode(): Int {
            var result = fileId.hashCode()
            result = 31 * result + name.hashCode()
            result = 31 * result + size.hashCode()
            result = 31 * result + mime.hashCode()
            result = 31 * result + checksum.contentHashCode()
            return result
        }
    }

    data class Chunk(
        val fileId: UUID,
        val index: Int,
        val data: ByteArray,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Chunk) return false
            return fileId == other.fileId &&
                index == other.index &&
                data.contentEquals(other.data)
        }

        override fun hashCode(): Int {
            var result = fileId.hashCode()
            result = 31 * result + index
            result = 31 * result + data.contentHashCode()
            return result
        }
    }

    data class Ack(val fileId: UUID, val index: Int, val status: Byte)

    /** Tag-prefixed frame parsed off the stream after the handshake. */
    sealed interface Frame {
        data class Header(val header: FileHeader) : Frame
        data class ChunkFrame(val chunk: Chunk) : Frame
        data class AckFrame(val ack: Ack) : Frame
        data class CancelFrame(val fileId: UUID) : Frame
    }

    // ---- Handshake ---------------------------------------------------------

    /** Reads the handshake — first frame on a fresh socket, no type tag. */
    fun readHandshake(input: DataInputStream): Handshake {
        val version = input.readShort()
        val deviceId = readUuid(input)
        val publicKey = ByteArray(32).also { input.readFully(it) }
        val capabilities = input.readUnsignedByte()
        return Handshake(version, deviceId, publicKey, capabilities)
    }

    fun writeHandshake(output: DataOutputStream, handshake: Handshake) {
        output.writeShort(handshake.version.toInt())
        writeUuid(output, handshake.deviceId)
        output.write(handshake.publicKey)
        output.writeByte(handshake.capabilities and 0xFF)
        output.flush()
    }

    // ---- Frames (post-handshake) -------------------------------------------

    /** Reads one tag-prefixed frame. Returns null on EOF. */
    fun readFrame(input: DataInputStream): Frame? {
        val tag = try {
            input.readByte()
        } catch (_: java.io.EOFException) {
            return null
        }
        return when (tag) {
            TYPE_FILE_HEADER -> Frame.Header(readFileHeaderBody(input))
            TYPE_CHUNK -> Frame.ChunkFrame(readChunkBody(input))
            TYPE_ACK -> Frame.AckFrame(readAckBody(input))
            TYPE_CANCEL -> Frame.CancelFrame(readUuid(input))
            else -> throw IllegalStateException("Unknown frame tag 0x${"%02x".format(tag)}")
        }
    }

    fun writeFileHeader(output: DataOutputStream, header: FileHeader) {
        output.writeByte(TYPE_FILE_HEADER.toInt())
        writeUuid(output, header.fileId)
        val nameBytes = header.name.toByteArray(StandardCharsets.UTF_8)
        output.writeShort(nameBytes.size)
        output.write(nameBytes)
        output.writeLong(header.size)
        val mimeBytes = header.mime.toByteArray(StandardCharsets.UTF_8)
        output.writeByte(mimeBytes.size)
        output.write(mimeBytes)
        output.write(header.checksum)
        output.flush()
    }

    fun writeChunk(output: DataOutputStream, chunk: Chunk) {
        output.writeByte(TYPE_CHUNK.toInt())
        writeUuid(output, chunk.fileId)
        output.writeInt(chunk.index)
        output.writeInt(chunk.data.size)
        output.write(chunk.data)
        // Caller flushes after a batch — flush per chunk would kill throughput.
    }

    fun writeAck(output: DataOutputStream, ack: Ack) {
        output.writeByte(TYPE_ACK.toInt())
        writeUuid(output, ack.fileId)
        output.writeInt(ack.index)
        output.writeByte(ack.status.toInt())
        output.flush()
    }

    fun writeCancel(output: DataOutputStream, fileId: UUID) {
        output.writeByte(TYPE_CANCEL.toInt())
        writeUuid(output, fileId)
        output.flush()
    }

    // ---- Internals ---------------------------------------------------------

    private fun readFileHeaderBody(input: DataInputStream): FileHeader {
        val fileId = readUuid(input)
        val nameLen = input.readUnsignedShort()
        val name = ByteArray(nameLen).also { input.readFully(it) }
            .toString(StandardCharsets.UTF_8)
        val size = input.readLong()
        val mimeLen = input.readUnsignedByte()
        val mime = ByteArray(mimeLen).also { input.readFully(it) }
            .toString(StandardCharsets.UTF_8)
        val checksum = ByteArray(32).also { input.readFully(it) }
        return FileHeader(fileId, name, size, mime, checksum)
    }

    private fun readChunkBody(input: DataInputStream): Chunk {
        val fileId = readUuid(input)
        val index = input.readInt()
        val size = input.readInt()
        require(size in 0..(8 * 1024 * 1024)) { "implausible chunk size: $size" }
        val data = ByteArray(size).also { input.readFully(it) }
        return Chunk(fileId, index, data)
    }

    private fun readAckBody(input: DataInputStream): Ack {
        val fileId = readUuid(input)
        val index = input.readInt()
        val status = input.readByte()
        return Ack(fileId, index, status)
    }

    private fun readUuid(input: DataInputStream): UUID {
        val msb = input.readLong()
        val lsb = input.readLong()
        return UUID(msb, lsb)
    }

    private fun writeUuid(output: DataOutputStream, uuid: UUID) {
        output.writeLong(uuid.mostSignificantBits)
        output.writeLong(uuid.leastSignificantBits)
    }

    /** Convenience: pack a 16-byte UUID into a ByteBuffer-shaped ByteArray. */
    fun uuidToBytes(uuid: UUID): ByteArray {
        val buf = ByteBuffer.allocate(16)
        buf.putLong(uuid.mostSignificantBits)
        buf.putLong(uuid.leastSignificantBits)
        return buf.array()
    }
}
